### ISSUE TITLE
Input data to predicted probability indexes

### ISSUE THREAD

**Issue by bjascob** [uds-lsv/TF-NNLM-TK#4]
Thanks for the great library for testing out language models.  I'd like to use this to predict the next word in a list of words and I'm having a bit of of trouble with indexing.  In order to get the softmax output of the model I've added to basic_rnn_models.py
```
with tf.name_scope('probs'):
    self.probs = tf.nn.softmax(self.logits)
```
This produces an output shape that is [num_batches, batch_size, seq_len].  The odd thing is that, looking at the test_data.data (1D array) vs test_data.input (3D array), it looks like the individual words fill "seq" (3rd array index) and then jump to the next epoch (1st array index) and then finally the batch index (2nd) increments.  Is there a reshape I can use to convert this softmax output back to a 1D array that corresponds to the order of the input (test_data.data)?  I don't think a simple flatten gets the order correct.

Give the ordering of things, I'm also curious about prediction dependency.  Does the lstm have a chance to "look back" only seq_len or it is seq_len * batch_size or seq_len * num_batches.  From looking at the code it looks like you're saving state across epochs so I'm thinking there's some level of dependency across epochs but not batches.  I'm interested your view on this.

**Comment by bjascob** [comment]
After some experimenting it looks like this is simpler than I expected.  The following appears to work...
```
# probs is [epoch, batch_size, seq_len]
probs = probs.swapaxes(0,1)         # probs is now [batch_size, epoch, seq_len].
probs = probs.reshape(probs.shape[0]*probs.shape[1]*probs.shape[2],-1)
```

**Comment by michael-aloys** [comment]
Hi,

nice that you already found a solution to your problem yourself.

To answer your second question: The model can look back seq_len * num_batches words (for the last word; in theory, not tacking into account problems with long distance dependencies). In detail:

To be able to improve runtime, mini-batch training is used. Given a corpus/dataset

a b c d e f g h i j k l

a sequence length of 3 and a batch size of 2 the data is split into

batch 1
`sequence1: a b  c
sequence2: g h i
`
batch 2
`sequence1: d e f
sequence2: j k l`

The sequence1 and sequence2 can then be trained in parallel. For the model to be able to look farther back then just within the sequence within one batch, the state of the sequence is transfered into the next batch (and that's the reason for the "strange" order of the sequences). So the state after processing c (in batch 1) is used as previous state when processing d (in batch 2). And the state after processing i is used as previous state when processing j.

Back-propagation stops at sequence length/the batch boundary, but for prediction the model uses the previous hidden state and therefore can look back much further.

I hope that answered your question.

**Comment by bjascob** [comment]
Thanks for the great explanation.  Looking at the code's various shaping operations it's difficult to figure out what's going on and why.  The above makes it clear how this is designed to work.
